consts not recognized in anonymous structs

### Describe the bug

This was previously identified in https://github.com/vlang/v/issues/19603, but I thought a separate, focused issue would be helpful.

Trying to use a const as a default in an anonymous struct parameter of a function results in a compiler error.

### Reproduction Steps

```
module main

const salutation = 'Good morning!'

// vfmt off
fn greeting(args struct {
        name string
        salut string = salutation
}) string {
        // vfmt on
        return 'Hi, ${args.name}! ${args.salut}'
}

fn main() {
        println(greeting(name: 'Flibby'))
}
```

### Expected Behavior

Console output would be "Hi, Flibby! Good morning!"

### Current Behavior

```
==================
/tmp/v_506/anon-struct-constant.4301051555783568569.tmp.c:12670:35: error: use of undeclared identifier 'salutation'
        .name = _SLIT("Flibby"),.salut = salutation,}));
                                         ^
1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

Add consts to the scope of anonymous structs.

### Additional Information/Context

Happens on Linux as well:
```
V full version: V 0.4.4 22cb9c5.ac2dcc2
OS: linux, PikaOS Linux 3
Processor: 12 cpus, 64bit, little endian, AMD Ryzen 5 5600 6-Core Processor

getwd: /home/flibby/MEGAsync/Development/vlang/anon-struct-const
vexe: /opt/vlang/v
vexe mtime: 2024-01-09 16:22:38

vroot: OK, value: /opt/vlang
VMODULES: OK, value: /home/flibby/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.40.1
Git vroot status: weekly.2023.38-838-gac2dcc2b
.git/config present: true

CC version: cc (Ubuntu 13.2.0-4ubuntu3) 13.2.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```

### V version

V 0.4.4 ac2dcc2

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.4 29511bc.ac2dcc2
OS: macos, macOS, 14.2.1, 23C71
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz

getwd: /opt/johann/Development/tmp/vlang/anon-struct-constant
vexe: /opt/vlang/v
vexe mtime: 2024-01-09 15:53:04

vroot: OK, value: /opt/vlang
VMODULES: OK, value: /Users/johann/.vmodules
VTMP: OK, value: /tmp/v_506

env VFLAGS: "-use-os-system-to-run"

Git version: git version 2.43.0
Git vroot status: weekly.2023.52-117-gac2dcc2b
.git/config present: true

CC version: Apple clang version 15.0.0 (clang-1500.1.0.2.5)
thirdparty/tcc status: thirdparty-macos-amd64 46662e20
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/struct.v
+++ b/vlib/v/checker/struct.v
@@ -792,9 +792,13 @@ or use an explicit `unsafe{ a[..] }`, if you do not want a copy of the slice.',
 							&& field.default_expr.info is ast.IdentFn {
 							c.expr(mut field.default_expr)
 						} else {
-							c.expected_type = field.typ
-							field.default_expr_typ = c.expr(mut field.default_expr)
-							info.fields[i].default_expr_typ = field.default_expr_typ
+							if const_field := c.table.global_scope.find_const('${field.default_expr}') {
+								info.fields[i].default_expr_typ = const_field.typ
+							} else if type_sym.info is ast.Struct && type_sym.info.is_anon {
+								c.expected_type = field.typ
+								field.default_expr_typ = c.expr(mut field.default_expr)
+								info.fields[i].default_expr_typ = field.default_expr_typ
+							}
 						}
 					}
 					continue

--- a/vlib/v/checker/struct.v
+++ b/vlib/v/checker/struct.v
@@ -792,9 +792,9 @@ or use an explicit `unsafe{ a[..] }`, if you do not want a copy of the slice.',
 							&& field.default_expr.info is ast.IdentFn {
 							c.expr(mut field.default_expr)
 						} else {
-							if const_field := c.table.global_scope.find_const('${field.default_expr}') {
-								info.fields[i].default_expr_typ = const_field.typ
-							}
+							c.expected_type = field.typ
+							field.default_expr_typ = c.expr(mut field.default_expr)
+							info.fields[i].default_expr_typ = field.default_expr_typ
 						}
 					}
 					continue

--- a/vlib/v/tests/anon_struct_with_default_expr_test.v
+++ b/vlib/v/tests/anon_struct_with_default_expr_test.v
@@ -13,3 +13,21 @@ fn test_anon_struct_with_default_expr() {
 	println(bar.anon.foofoo.name)
 	assert bar.anon.foofoo.name == 'foofoo'
 }
+
+// for issue 20452
+// phenomenon:
+// cgen generates incorrect code when the default value of an anonymous struct field is the const variable.
+const hello = 'hello'
+
+// vfmt off
+fn method_passed_an_anon_struct_arg(arg struct {
+		name string
+		greeting string = hello
+	}) string {
+	return '${arg.greeting} ${arg.name}!'
+}
+// vfmt on
+
+fn test_anon_struct_with_const_default_expr() {
+	assert method_passed_an_anon_struct_arg(name: 'world') == 'hello world!'
+}

